File: 169-majority-element/169-majority_element.ts
/*
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times.
You may assume that the majority element always exists in the array.


*/
function majorityElement(nums: number[]): number {
    const n = nums.length

    let count = {}
    for (let i = 0; i < n; i++) {
        if (!(nums[i] in count)) {
            count[nums[i]] = 1
        } else {
            count[nums[i]]++
        }
        if (count[nums[i]] > n / 2) {
            return nums[i]
        }
    }

    return -1
};

File: 169-majority-element/169-majority_element.py
from collections import defaultdict


def majorityElement(nums: list[int]):
    """
    Given an array nums of size n, return the majority element.

    The majority element is the element that appears more than ⌊n / 2⌋ times. 
    You may assume that the majority element always exists in the array.
    """
    count = defaultdict(int)
    n = len(nums)
    for num in nums:
        count[num] += 1
        if count[num] > n//2:
            return num


if __name__ == "__main__":
    print(majorityElement([3,2,3]))
    print(majorityElement([2,2,1,1,1,2,2]))
    print(majorityElement([1,1]))
